### DISCUSSION TITLE
Custom U8g2 font

### DISCUSSION THREAD

**Discussion by vzahradnik** [davetcc/tcMenu#372] [Q&A]
I made a custom U8g2 font with extra glyphs to support my unicode characters. It work correctly with TcMenu when I update the file `tcMenuU8g2.cpp`.

In this file I include my custom font header and then modify the following function:

```cpp
const uint8_t* safeGetFont(const void* fnt) {
    if(fnt) return static_cast<const uint8_t *>(fnt);
    return u8g2_font_6x10_tf;
}
```

However, the file `tcMenuU8g2.cpp` is overwritten every time I re-generate my code from the Designer. Is there a better way how to define my custom font? I tried to specify it in the Designer but so far I was not successful. The generated code still uses the original font, not my custom one.

Thanks!

**Comment by davetcc** [comment]
I mean in the designer you should be able to use the following method for U8G2 pre-packaged fonts, which is basically opening code generator dialog, select the "set font" button, choose U8G2/static, and enter the font. However, you probably need it to work more like an adafruit font where you include the header, and I think it would just work if you chose the Adafruit local to project option as it would then include the header file (assuming a header file of the same name as the font).

![image](https://github.com/davetcc/tcMenu/assets/12195465/afb21b6a-9488-49fe-aef5-a09705186c3d)

Another option is that you manually setup the font in the theme header file, and set the font to default in the designer. Theme files are safe to edit, as they only get regenerated if you delete the theme file.

**Reply by vzahradnik** [reply]
I don't mind to use any supported option. But I think that this option where I specify the variable doesn't work properly. The generated code, even after this change, used the original font.

I will do some more testing to get more information.

**Reply by davetcc** [reply]
I think there is even an example that has a u8g2 custom font. You should be able to look at that for guidance.

**Comment by vzahradnik** [comment]
@davetcc I tried both your solutions. However they don't work for my usecase.

1. Static Font (U8G2 displays) works only for built-in U8G2 fonts. My custom header is not included. You anticipated this case.
2. Adafruit font local to project. It allows to include custom header but it defines a font variable of the following type:

```cpp
extern const GFXfont u8g2_font_6x10_t_complete;
```

Instead, my font is defined like this:
```cpp
extern const uint8_t u8g2_font_6x10_t_complete[] U8G2_FONT_SECTION("u8g2_font_6x10_t_complete");
```

My usecase is therefore not covered. I can't change font structure to Adafruit `GFXfont`. I suggest to add a new option to include custom header. Also, when custom header is included, TcMenu should not generate any font variables (fields in the form would still be needed to tell TcMenu the variable name to use). This should give us the flexibility to cover all future cases.

To sum up:
1. Add new option to TcMenu designer to define header path, like "fonts/acdu_fonts.h" or <fonts/acdu_fonts.h>. This would have the highest priority and should work for all usecases, either Adafruit, U8G2 or possibly others.
2. Header contains font variable declaration. Responsibility to properly define this variable is on users.
3. Generator includes this header in `menu.h`
4. Variable is used in `installMonoBorderedTheme(renderer, MenuFontDef(...)`

By the way, how do you differentiate between different font structs? `MenuFontDef` assumes every font is `const void *data`.
